README.md
# docker-machine-ucloud

docker-machine-ucloud is a plugin of docker-machine for UCloud. It requires Docker Machine's version is greater than v0.5.0-rc1.

（[中文文档](README-zh.md)） 

# Install and Run

You could install run docker-machine-driver-ucloud as following steps.

First, docker-machine v0.5.2 is required, documentation for how to install `docker-machine`
[is available here](https://github.com/docker/machine/releases/tag/v0.5.2#Installation).

or you can install `docker-machine` from source code by running these commands
```
$ go get github.com/docker/machine
$ cd $GOPATH/src/github.com/docker/machine
$ make build
```

Then, you could install `docker-machine-ucloud` driver in the $GOPATH and add $GOPATH/bin to the $PATH env. 

```
go get github.com/ucloud/docker-machine-ucloud
cd $GOPATH/src/github.com/ucloud/docker-machine-ucloud
make
make install
```

Now, you can run `docker-machine create --help -d ucloud` to see how to create a UCloud machine. Public and private keys of UCloud API
are needed to create machine. Both options and environment variable are available to set that:

```
$ export UCLOUD_PUBLIC_KEY=<public-key-of-ucloud-api>
$ export UCLOUD_PUBLIC_KEY=<private-key-of-ucloud-api>
$ docker-machine create -d ucloud <machine-name>
```
or  run 

```
$ docker-machine create -d ucloud --ucloud-public-key=<public-key> --ucloud-private-key=<private-key> <machine-name>
```

for example,

```
$ ./docker-machine create -d ucloud ucloud-machine
Running pre-create checks...
Creating machine...
Waiting for machine to be running, this may take a few minutes...
Machine is running, waiting for SSH to be available...
Detecting operating system of created instance...
Provisioning created instance...
Copying certs to the local machine directory...
Copying certs to the remote machine...
Setting Docker configuration on the remote daemon...
To see how to connect Docker to this machine, run: docker-machine_darwin-amd64 env ucloud-machine
```

After run this command, a machine with name `ucloud-machine` is created.

```
$ ./docker-machine ls
NAME             ACTIVE   DRIVER   STATE     URL                        SWARM
ucloud-machine   -        ucloud   Running   tcp://123.59.66.163:2376
```



ChangeLog.md
# ChangeLog

# 0.5.2 (2015-12-03)
1. 支持swarm集群的创建，swarm master的默认端口3376
2. 支持主机配置的定制，增加 —ucloud-cpu-cores, —ucloud-disk-space, —ucloud-memory-size。默认CPU: 1核，Memory: 1G, Disk: 20G 
3. 支持付费类型的定制，增加 —ucloud-charge-type 默认Month, 支持Year, Month, Dynamic, Trial
4. 调用的API请求设置User-Agent 为 docker-machine/v0.5.2

# 0.5.0-rc3 (2015-11-03)
1. 支持UHost的创建


client_test.go
package ucloud


import (
	"testing"
)

func TestCreateUNet(t *testing.T) {
	d := Driver{
		Region: "cn-north-03",
		UhostID: "uhost-ygtk1p",
	}

	err := d.createUNet()
	if  err != nil {
		t.Errorf("create UNet failed:%s", err)
	}
}


ucloud_test.go
package ucloud


utils.go
package ucloud

import (
	"errors"
	"math/rand"
	"time"
)

var (
	errInvalidRegion = errors.New("invalid region specified")
)

var regions = []string{
	"cn-north-01",
	"cn-north-02",
	"cn-north-03",
	"cn-east-01",
	"cn-south-01",
	"cn-south-02",
	"hk-01",
	"us-west-01",
}

func validateUCloudRegion(region string) (string, error) {
	for _, v := range regions {
		if v == region {
			return region, nil
		}
	}

	return "", errInvalidRegion
}

func validPort(port int) bool {
	if port < 1 || port > 65535 {
		return false
	}

	return true
}
func generateRandomPassword(n int) string {
	rand.Seed(time.Now().UnixNano())
	var letters = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ~!@#$%^&*()_+}{:?><")

	b := make([]rune, n)
	for i := range b {
		b[i] = letters[rand.Intn(len(letters))]
	}

	return string(b)
}


client.go
package ucloud

import (
	"encoding/base64"
	"fmt"
	"io/ioutil"
	"strings"

	"github.com/docker/machine/libmachine/log"
	"github.com/docker/machine/libmachine/mcnutils"
	"github.com/docker/machine/libmachine/ssh"
	"github.com/docker/machine/version"
	"github.com/ucloud/ucloud-sdk-go/service/uhost"
	"github.com/ucloud/ucloud-sdk-go/service/unet"
	"github.com/ucloud/ucloud-sdk-go/ucloud"
	"github.com/ucloud/ucloud-sdk-go/ucloud/auth"
)

var (
	hostsvc *uhost.UHost
	unetsvc *unet.UNet
)

func (d *Driver) newConfig() *ucloud.Config {
	return &ucloud.Config{
		Credentials: &auth.KeyPair{
			PublicKey:  d.PublicKey,
			PrivateKey: d.PrivateKey,
		},
		Region: d.Region,
		HTTPHeader: map[string]string{
			"User-Agent": "docker-machine/" + version.Version,
		},
	}
}

func (d *Driver) getUHostService() *uhost.UHost {

	if hostsvc != nil {
		return hostsvc
	}
	hostsvc = uhost.New(d.newConfig())

	return hostsvc
}

func (d *Driver) getUNetService() *unet.UNet {

	if unetsvc != nil {
		return unetsvc
	}
	unetsvc = unet.New(d.newConfig())

	return unetsvc
}

func (d *Driver) createUHost() error {
	password := strings.Replace(base64.StdEncoding.EncodeToString([]byte(d.Password)), "=", "", -1)

	createUhostParams := uhost.CreateUHostInstanceParams{

		Region:     d.Region,
		ImageId:    d.ImageId,
		LoginMode:  "Password",
		Password:   password,
		CPU:        d.CPU,
		Memory:     d.Memory,
		DiskSpace:  d.DiskSpace,
		Name:       d.MachineName,
		ChargeType: d.ChargeType,
		Quantity:   1,
		Count:      1,
	}

	resp, err := d.getUHostService().CreateUHostInstance(&createUhostParams)
	if err != nil {
		return err
	}

	if resp == nil {
		return fmt.Errorf("response is empty")
	}

	if len(resp.UHostIds) == 0 {
		return fmt.Errorf("UHostIds is empty")
	}
	d.UhostID = resp.UHostIds[0]

	return nil
}

func (d *Driver) startUHost() error {
	startUhostParams := uhost.StartUHostInstanceParams{
		Region:  d.Region,
		UHostId: d.UhostID,
	}
	_, err := d.getUHostService().StartUHostInstance(&startUhostParams)
	if err != nil {
		return err
	}

	return nil
}

func (d *Driver) killUHost() error {
	killUHostParams := uhost.PoweroffUHostInstanceParams{
		Region:  d.Region,
		UHostId: d.UhostID,
	}

	_, err := d.getUHostService().PoweroffUHostInstance(&killUHostParams)
	if err != nil {
		return err
	}

	return nil
}

func (d *Driver) rebootUHost() error {

	killUHostParams := uhost.PoweroffUHostInstanceParams{
		Region:  d.Region,
		UHostId: d.UhostID,
	}

	_, err := d.getUHostService().PoweroffUHostInstance(&killUHostParams)
	if err != nil {
		return err
	}
	return nil
}

func (d *Driver) terminateUHost() error {

	terminateUHostParams := uhost.TerminateUHostInstanceParams{
		Region:  d.Region,
		UHostId: d.UhostID,
	}

	_, err := d.getUHostService().TerminateUHostInstance(&terminateUHostParams)
	if err != nil {
		return err
	}

	// TODO: remove EIP and security group etc.
	return nil
}

func (d *Driver) stopUHost() error {
	stopUhostParams := uhost.StopUHostInstanceParams{
		Region:  d.Region,
		UHostId: d.UhostID,
	}

	_, err := d.getUHostService().StopUHostInstance(&stopUhostParams)
	if err != nil {
		return err
	}

	return nil
}

type UHostDetail struct {
	region string
	hostID string

	state            string
	publicIPAddress  string
	privateIPAddress string
	cpu              int
	memory           int
}

func (d *Driver) getHostDescription() (*UHostDetail, error) {

	describeParams := uhost.DescribeUHostInstanceParams{
		Region:   d.Region,
		UHostIds: []string{d.UhostID},
		Offset:   0,
		Limit:    10,
	}

	resp, err := d.getUHostService().DescribeUHostInstance(&describeParams)
	if err != nil {
		return nil, err
	}

	if len(resp.UHostSet) == 0 {
		return nil, fmt.Errorf("UHost is not exist.")
	}

	if len(resp.UHostSet[0].IPSet) == 0 {
		return nil, fmt.Errorf("IPSet is not exist")
	}

	var publicIpAddress string
	var privateIPAddress string
	for _, ip := range resp.UHostSet[0].IPSet {
		switch ip.Type {
		case "Private":
			privateIPAddress = ip.IP
		case "Bgp":
			publicIpAddress = ip.IP
		}
	}

	d.CPU = resp.UHostSet[0].CPU
	d.Memory = resp.UHostSet[0].Memory

	return &UHostDetail{
		region:           d.Region,
		hostID:           resp.UHostSet[0].UHostId,
		state:            resp.UHostSet[0].State,
		publicIPAddress:  publicIpAddress,
		privateIPAddress: privateIPAddress,
		cpu:              resp.UHostSet[0].CPU,
		memory:           resp.UHostSet[0].Memory,
	}, nil
}

// createUNet create network for uhost
func (d *Driver) createUNet() error {
	if err := d.configureIPAddress(); err != nil {
		return fmt.Errorf("configure IPAddress error:%s", err)
	}

	if err := d.configureSecurityGroup(); err != nil {
		return fmt.Errorf("configure security group error:%s", err)
	}

	return nil
}

// createKeyPair create keypair for ssh to docker-machine
func (d *Driver) createKeyPair() error {
	log.Debugf("SSH key path:%s", d.GetSSHKeyPath())

	if err := ssh.GenerateSSHKey(d.GetSSHKeyPath()); err != nil {
		return err
	}

	return nil
}

func (d *Driver) waitForSSHFunc(client ssh.Client, command string) func() bool {
	return func() bool {
		_, err := client.Output(command)
		if err == nil {
			return true
		}
		return false
	}
}

// uploadKeyPair upload the public key to docker-machine
func (d *Driver) uploadKeyPair() error {

	ipAddr, err := d.GetIP()
	if err != nil {
		return err
	}

	port, _ := d.GetSSHPort()
	auth := ssh.Auth{
		Passwords: []string{d.Password},
	}

	ssh.SetDefaultClient(ssh.Native)
	sshClient, err := ssh.NewClient(d.GetSSHUsername(), ipAddr, port, &auth)
	if err != nil {
		return err
	}
	d.waitForSSHFunc(sshClient, "exit 0")

	publicKey, err := ioutil.ReadFile(d.GetSSHKeyPath() + ".pub")
	if err != nil {
		return err
	}

	command := fmt.Sprintf("mkdir -p ~/.ssh; echo '%s' > ~/.ssh/authorized_keys", string(publicKey))
	log.Debugf("Upload the public key with command: %s", command)

	output, err := sshClient.Output(command)
	if err != nil {
		log.Debugf("Upload command err, output: %v: %s", err, output)
		return err
	}

	return nil
}

func (d *Driver) configureIPAddress() error {

	// create an EIP and bind it to host
	if !d.PrivateIPOnly {
		createEIPParams := unet.AllocateEIPParams{
			Region:       d.Region,
			OperatorName: "Bgp",
			Bandwidth:    2,
			ChargeType:   "Dynamic",
			Quantity:     1,
		}

		resp, err := d.getUNetService().AllocateEIP(&createEIPParams)
		if err != nil {
			return fmt.Errorf("Allocate EIP failed:%s", err)
		}
		log.Debug(resp)

		if len(*resp.EIPSet) == 0 {
			return fmt.Errorf("EIP is empty")
		}
		eipId := (*resp.EIPSet)[0].EIPId
		if len(*(*resp.EIPSet)[0].EIPAddr) == 0 {
			return fmt.Errorf("IP Address is empty")
		}
		d.IPAddress = (*(*resp.EIPSet)[0].EIPAddr)[0].IP

		bindHostParams := unet.BindEIPParams{
			Region:       d.Region,
			EIPId:        eipId,
			ResourceType: "uhost",
			ResourceId:   d.UhostID,
		}

		bindEIPResp, err := d.getUNetService().BindEIP(&bindHostParams)
		if err != nil {
			return fmt.Errorf("Bind EIP failed:%s", err)
		}
		log.Debug(bindEIPResp)
	} else {
		hostDetails, err := d.getHostDescription()
		if err != nil {
			return fmt.Errorf("get host detail failed: %s", err)
		}
		d.IPAddress = hostDetails.publicIPAddress
		d.PrivateIPAddress = hostDetails.privateIPAddress
	}

	return nil
}

func (d *Driver) getSecurityGroup(name string) (int, error) {
	log.Debugf("get security group for group:%s", name)
	describeSecurityGroupsParams := unet.DescribeSecurityGroupParams{
		Region: d.Region,
	}
	describeSecurityGroupsResp, err := d.getUNetService().DescribeSecurityGroup(&describeSecurityGroupsParams)
	if err != nil {
		return 0, fmt.Errorf("get security groups failed:%s", err)
	}

	if len(describeSecurityGroupsResp.DataSet) == 0 {
		return 0, fmt.Errorf("security groups is empty")
	}

	for _, groups := range describeSecurityGroupsResp.DataSet {
		log.Debugf("name:%s, group id:%d", groups.GroupName, groups.GroupId)
		if groups.GroupName == name {
			log.Debugf("groups:%+v", groups)
			return groups.GroupId, nil
		}
	}

	return 0, fmt.Errorf("group:%s is not exist", name)
}

func (d *Driver) securityGroupAvailableFunc(name string) func() bool {
	return func() bool {
		_, err := d.getSecurityGroup(name)
		if err == nil {
			return true
		}
		return false
	}
}

func (d *Driver) configureSecurityGroup() error {
	var groupId int
	groupId, err := d.getSecurityGroup(d.SecurityGroupName)
	if err != nil {
		log.Debugf("get security group error:%s", err)
	}
	log.Debugf("groupId:%d", groupId)
	if groupId == 0 {
		log.Infof("security group is not found, create a new one")
		rule := []string{"TCP|22|0.0.0.0/0|ACCEPT|50",
			"TCP|3389|0.0.0.0/0|ACCEPT|50",
			"TCP|2376|0.0.0.0/0|ACCEPT|50",
		}
		if d.SwarmMaster && validPort(swarmPort) {
			swarmRule := fmt.Sprintf("TCP|%d|0.0.0.0/0|ACCEPT|50", swarmPort)
			rule = append(rule, swarmRule)
		}

		securityGroupParams := unet.CreateSecurityGroupParams{
			Region:      d.Region,
			GroupName:   "docker-machine",
			Description: "docker machine to open 2379 and 22 port of tcp",
			Rule:        rule,
		}
		_, err := d.getUNetService().CreateSecurityGroup(&securityGroupParams)
		if err != nil {
			return fmt.Errorf("create security group failed:%s", err)
		}

		log.Debug("waiting for security group to become avaliable")
		if err := mcnutils.WaitFor(d.securityGroupAvailableFunc(d.SecurityGroupName)); err != nil {
			return err
		}
		groupId, err = d.getSecurityGroup(d.SecurityGroupName)
	}
	d.SecurityGroupId = groupId

	grantSecurityGroupParams := unet.GrantSecurityGroupParams{
		Region:       d.Region,
		GroupId:      groupId,
		ResourceType: "uhost",
		ResourceId:   d.UhostID,
	}
	log.Debugf("grant security group(%d) to uhost(%s)", groupId, d.UhostID)
	_, err = d.getUNetService().GrantSecurityGroup(&grantSecurityGroupParams)
	if err != nil {
		return fmt.Errorf("grant security group failed:%s", err)
	}

	return nil
}


ucloud.go
package ucloud

import (
	"fmt"
	"net/url"
	"strconv"
	"strings"
	"time"

	"github.com/docker/machine/libmachine/drivers"
	"github.com/docker/machine/libmachine/log"
	"github.com/docker/machine/libmachine/mcnflag"
	"github.com/docker/machine/libmachine/mcnutils"
	"github.com/docker/machine/libmachine/state"
)

type Driver struct {
	*drivers.BaseDriver

	PublicKey  string
	PrivateKey string
	Region     string
	ImageId    string
	Password   string
	UhostID    string

	CPU        int
	Memory     int
	DiskSpace  int
	ChargeType string

	PrivateIPOnly     bool
	PrivateIPAddress  string
	SecurityGroupId   int
	SecurityGroupName string
}

const (
	defaultTimeout    = 1 * time.Second
	defaultCPU        = 1
	defaultMemory     = 1024
	defaultDiskSpace  = 20
	defaultRegion     = "cn-north-03"
	defaultChargeType = "Month"
	defaultRetries    = 10
	defaultImageId    = "uimage-5yt2b0" // we use CentOS 7.0 default
)

var (
	swarmPort = 3376
)

func NewDriver(hostName, artifactPath string) *Driver {
	return &Driver{
		BaseDriver: &drivers.BaseDriver{
			MachineName: hostName,
			//ArtifactPath: artifactPath,
		},
		Region:    defaultRegion,
		Memory:    defaultMemory,
		CPU:       defaultCPU,
		DiskSpace: defaultDiskSpace,
	}
}

func (d *Driver) GetCreateFlags() []mcnflag.Flag {
	return []mcnflag.Flag{

		mcnflag.StringFlag{
			Name:   "ucloud-public-key",
			Usage:  "UCloud Public Key",
			Value:  "",
			EnvVar: "UCLOUD_PUBLIC_KEY",
		},
		mcnflag.StringFlag{
			Name:   "ucloud-private-key",
			Usage:  "UCloud Private Key",
			Value:  "",
			EnvVar: "UCLOUD_PRIVATE_KEY",
		},
		mcnflag.StringFlag{
			Name:  "ucloud-imageid",
			Usage: "UHost image id",
			Value: "",
		},
		mcnflag.StringFlag{
			Name:   "ucloud-region",
			Usage:  "Region of ucloud idc",
			Value:  "cn-north-03",
			EnvVar: "UCLOUD_REGION",
		},
		mcnflag.StringFlag{
			Name:  "ucloud-ssh-user",
			Usage: "SSH user",
			Value: "root",
		},
		mcnflag.IntFlag{
			Name:  "ucloud-cpu-core",
			Usage: "Number of CPU cores,default is 1",
			Value: defaultCPU,
		},
		mcnflag.IntFlag{
			Name:  "ucloud-memory-size",
			Usage: "Size of memory, unit(MB), default 1024M",
			Value: defaultMemory,
		},
		mcnflag.IntFlag{
			Name:  "ucloud-disk-space",
			Usage: "Disk size, unit(GB),default is 20G",
			Value: defaultDiskSpace,
		},
		mcnflag.StringFlag{
			Name:  "ucloud-charge-type",
			Usage: "How to pay for, you can chose from (Year,Month,Dynamic,Trial),default is Month",
			Value: defaultChargeType,
		},
		mcnflag.IntFlag{
			Name:  "ucloud-ssh-port",
			Usage: "SSH port",
			Value: 22,
		},
		mcnflag.StringFlag{
			Name:  "ucloud-user-password",
			Usage: "Password of ucloud user, random password will be used if not set",
			Value: "",
		},
		mcnflag.BoolFlag{
			Name:  "ucloud-private-address-only",
			Usage: "Only use a private IP address",
		},
		mcnflag.StringFlag{
			Name:  "ucloud-security-group",
			Usage: "UCloud security group",
			Value: "docker-machine",
		},
	}
}

func (d *Driver) DriverName() string {
	return "ucloud"
}

func (d *Driver) GetSSHHostname() (string, error) {
	return d.GetIP()
}

func (d *Driver) GetSSHUsername() string {
	if d.SSHUser == "" {
		d.SSHUser = "root"
	}

	return d.SSHUser
}

func (d *Driver) setDefaultConfig() {
	d.Memory = defaultMemory
	d.CPU = defaultCPU
	d.ChargeType = defaultChargeType
	d.DiskSpace = defaultDiskSpace
	d.Region = defaultRegion
	d.ImageId = defaultImageId
}

func (d *Driver) isSwarmMaster() bool {
	return d.SwarmMaster
}

func (d *Driver) SetConfigFromFlags(flags drivers.DriverOptions) error {
	d.setDefaultConfig()
	region, err := validateUCloudRegion(flags.String("ucloud-region"))
	if err != nil {
		return err
	}
	d.Region = region

	d.PublicKey = flags.String("ucloud-public-key")
	if d.PublicKey == "" {
		return fmt.Errorf("ucloud driver requires the --ucloud-public-key option")
	}
	log.Debugf("ucloud public key: %s", d.PublicKey)

	d.PrivateKey = flags.String("ucloud-private-key")
	if d.PrivateKey == "" {
		return fmt.Errorf("ucloud driver requires the --ucloud-private-key option")
	}
	log.Debugf("ucloud private key: %s", d.PrivateKey)

	image := flags.String("ucloud-imageid")
	if len(image) == 0 {
		image = defaultImageId
	}
	d.ImageId = image
	d.CPU = flags.Int("ucloud-cpu-core")
	d.Memory = flags.Int("ucloud-memory-size")
	d.DiskSpace = flags.Int("ucloud-disk-space")
	d.ChargeType = flags.String("ucloud-charge-type")

	d.PrivateIPOnly = flags.Bool("ucloud-private-address-only")
	d.SecurityGroupName = flags.String("ucloud-security-group")

	d.SSHUser = strings.ToLower(flags.String("ucloud-ssh-user"))
	if d.SSHUser == "" {
		d.SSHUser = "root"
	}
	d.Password = flags.String("ucloud-user-password")
	d.SSHPort = flags.Int("ucloud-ssh-port")

	d.SwarmMaster = flags.Bool("swarm-master")
	d.SwarmHost = flags.String("swarm-host")
	d.SwarmDiscovery = flags.String("swarm-discovery")

	if d.isSwarmMaster() {
		u, err := url.Parse(d.SwarmHost)
		if err != nil {
			return fmt.Errorf("error parsing swarm host: %s", err)
		}

		parts := strings.Split(u.Host, ":")
		port, err := strconv.Atoi(parts[1])
		if err != nil {
			return err
		}

		swarmPort = port
	}

	return nil
}

func (d *Driver) PreCreateCheck() error {
	if d.CPU < 1 || d.CPU > 16 {
		return fmt.Errorf("CPU cores must be in set of (1,2,4,8,16)")
	}
	if d.Memory < 1024 || d.Memory > 65536 {
		return fmt.Errorf("Memory must be in range of [2048, 65536) with step of 2048MB, you can set 1024 in beijing-BGP-C")
	}
	if d.DiskSpace > 1000 {
		return fmt.Errorf("Disk space must in range of [0, 1000) with step of 10GB")
	}
	return nil
}

func (d *Driver) Create() error {
	log.Infof("Create UHost instance...")

	if d.Password == "" {
		d.Password = generateRandomPassword(16)
		log.Infof("password is not set, we use the random password instead, password:%s", d.Password)
	}

	// create keypair
	log.Infof("Creating key pair for instances...")
	if err := d.createKeyPair(); err != nil {
		return fmt.Errorf("unable to create key pair: %s", err)
	}

	// create uhost instance
	log.Infof("Creating uhost instance...")
	if err := d.createUHost(); err != nil {
		return fmt.Errorf("create UHost failed:%s", err)
	}

	// waiting for creating successful
	if err := mcnutils.WaitForSpecific(drivers.MachineInState(d, state.Running), 120, 3*time.Second); err != nil {
		return fmt.Errorf("wait for machine running failed: %s", err)
	}

	// create networks, like private ip, eip, and security group
	log.Infof("Creating networks...")
	//TODO: user the exist eip and security group to configure network
	if err := d.createUNet(); err != nil {
		return fmt.Errorf("create networks failed:%s", err)
	}

	// upload keypair
	if err := d.uploadKeyPair(); err != nil {
		return fmt.Errorf("upload keypair failed:%s", err)
	}

	return nil
}

func (d *Driver) GetURL() (string, error) {
	ip, err := d.GetIP()
	if err != nil {
		return "", err
	}
	if ip == "" {
		return "", nil
	}

	return fmt.Sprintf("tcp://%s:2376", ip), nil
}

func (d *Driver) GetIP() (string, error) {
	if !d.PrivateIPOnly && d.IPAddress == "" {
		return "", fmt.Errorf("IP address is not set")
	}
	if d.PrivateIPOnly && d.PrivateIPAddress == "" {
		return "", fmt.Errorf("Private address is not set")
	}

	s, err := d.GetState()
	if err != nil {
		return "", err
	}
	if s != state.Running {
		return "", drivers.ErrHostIsNotRunning
	}

	if d.PrivateIPOnly {
		return d.PrivateIPAddress, nil
	}

	return d.IPAddress, nil
}

func (d *Driver) GetState() (state.State, error) {
	log.Debugf("Get Machine State")
	if d.UhostID == "" || d.Region == "" {
		return state.None, fmt.Errorf("region or uhost is empty")
	}

	details, err := d.getHostDescription()
	if err != nil {
		return state.None, err
	}

	var st state.State
	if details != nil && details.state != "" {
		switch details.state {
		case "Initializing", "Starting", "Rebooting":
			st = state.Starting
		case "Running":
			st = state.Running
		case "Stopped":
			st = state.Stopped
		case "Stopping":
			st = state.Stopping
		case "Install Fail":
			st = state.Error
		default:
			st = state.None
		}
	}

	return st, nil
}

func (d *Driver) Start() error {
	log.Info("Start UHost...")
	if err := d.startUHost(); err != nil {
		return fmt.Errorf("Cannot start Machine:%s, with UHost: %s.", d.MachineName, d.UhostID)
	}

	return nil
}

func (d *Driver) Stop() error {
	log.Info("Stop UHost...")
	if len(d.UhostID) == 0 {
		return fmt.Errorf("UHost is not exist for Machine: %s", d.MachineName)
	}

	if err := d.stopUHost(); err != nil {
		return fmt.Errorf("Cannot start Machine:%s, with UHost: %s.", d.MachineName, d.UhostID)
	}

	return nil
}

func (d *Driver) Remove() error {
	log.Debug("Removing...")
	if err := d.terminateUHost(); err != nil {
		return fmt.Errorf("Unable to terminate the UHost instance: %s", err)
	}

	//TODO: any cleanup ?
	return nil
}

func (d *Driver) Restart() error {
	log.Debug("Restarting...")
	if err := d.rebootUHost(); err != nil {
		return fmt.Errorf("Unable to restart the UHost instance: %s", err)
	}

	return nil
}

func (d *Driver) Kill() error {
	log.Debug("Killing...")
	if err := d.killUHost(); err != nil {
		return fmt.Errorf("Unable to kill the UHost instance: %s", err)
	}

	return nil
}
